README.md
# YTB [![Build Status](https://travis-ci.org/raasakh/ytbnode.svg?branch=master)](https://travis-ci.org/raasakh/ytbnode) [![](https://images.microbadger.com/badges/version/raasakh/ytb-testnet.svg)]

In the master branch there is a code with functions that is under development. The latest release for each network can be found in the [Releases section](https://github.com/raasakh/ytbnode/releases), you can switch to the corresponding tag and build the application.

For further information please refer the official [documentation](https://docs.ytb.one).




# YTB Node Installation

```
sudo apt-get update 
sudo apt-get upgrade

sudo apt install apt-transport-https
sudo apt install software-properties-common
sudo add-apt-repository -y ppa:webupd8team/java
sudo apt-get update
sudo apt-get -y install oracle-java8-installer

java -version
sudo apt install default-jre
wget https://github.com/raasakh/ytbnode/releases/download/v1.0.0/ytb-node_1.0.0.deb
sudo dpkg -i ytb-node_1.0.0.deb
sudo nano /usr/share/ytb/conf/ytb.conf
```

```
miner {
   # Enable/disable block generation
   enable = yes

   # Required number of connections (both incoming and outgoing) to attempt block generation. Setting this value to 0
   # enables "off-line generation".
   quorum = 2

   # Enable block generation only in the last block if not older the given period of time
   interval-after-last-block-then-generation-is-allowed = 1d

   # Interval between microblocks
   micro-block-interval = 5s

   # Mininmum time interval between blocks
   minimal-block-generation-offset = 1001ms

   # Max amount of transactions in key block
   max-transactions-in-key-block = 0

   # Max amount of transactions in micro block
   max-transactions-in-micro-block = 255

   # Miner references the best microblock which is at least this age
   min-micro-block-age = 6s
 }
```

sudo systemctl enable ytb.service

sudo systemctl restart ytb.service

sudo systemctl start ytb.service

sudo systemctl stop ytb.service

or sudo service ytb restart

sudo service ytb start

sudo service ytb stop

**juornal** (exit ctrl+o)

sudo journalctl -u ytb.service -f



## Compiling Packages from source

It is only possible to create deb and fat jar packages.

### Install SBT (Scala Build Tool)

For Ubuntu/Debian:

```
echo "deb https://dl.bintray.com/sbt/debian /" | sudo tee -a /etc/apt/sources.list.d/sbt.list
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 2EE0EA64E40A89B84B2DF73499E82A75642AC823
sudo apt-get update
sudo apt-get install sbt
```

You can install sbt on Mac OS X using Homebrew.

### Compile Genesis

```
sbt "test:runMain tools.GenesisBlockGenerator genesis.conf"
```

Added genesis data to ytbnet.conf

### Running

sbt "run ytbnet.conf"


### Create Package

Clone this repo and execute

```
sbt packageAll
```

.deb and .jar packages will be in /package folder. To build testnet packages use

```
sbt -Dnetwork=testnet packageAll
```




# Running Tests

`sbt test`

**Note**

If you prefer to work with _SBT_ in the interactive mode, open it with settings:
```bash
SBT_OPTS="${SBT_OPTS} -Xms512M -Xmx1536M -Xss1M -XX:+CMSClassUnloadingEnabled" sbt
```

For Java9 it should be:
```
SBT_OPTS="${SBT_OPTS} -Xms512M -Xmx1536M -Xss1M -XX:+CMSClassUnloadingEnabled --add-modules=java.xml.bind --add-exports java.base/jdk.internal.ref=ALL-UNNAMED" sbt
```

to solve the `Metaspace error` problem.




# Running Integration Tests


## TL;DR

 * Make sure you have [Docker](https://www.docker.com/get-docker) and SBT.
 * `sbt it/test`


## Customizing Tests

By default, `it/test` will do the following:
* Build a container image with the fat jar and a [template.conf](src/it/resources/template.conf). The newly-built image
  will be registered with the local Docker daemon. This image is built with [sbt-docker](https://github.com/marcuslonnberg/sbt-docker)
  plugin.
* Run the test suites from `src/it/scala`, passing docker image ID via `docker.imageId` system property.

### Logging

By [default](src/main/resources/logback.xml) all logs are written to the STDOUT. If you want to write logs, for example,
to JSON files, you should define your own logging configuration and specify a path to it in `conf/application.ini`:

```
-Dlogback.configurationFile=/path/to/your/logback.xml
```

### Debugging

Integration tests run in a forked JVM. To debug test suite code launched by SBT, you will need to add remote debug
options to `javaOptions` in `IntegrationTest` configuration:

```scala
javaOptions in IntegrationTest += "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005"
```

Debugging a node inside a container is a little more complicated: you will need to modify the `MIR_OPTS` environment
variable before starting a container.




# Collecting performance metrics

**Note**: all required tools will be installed though [Docker](https://docs.docker.com) for simplicity.

1. Install [Graphite](https://graphite.readthedocs.io/en/latest/install.html#docker), a service for collecting metrics.
2. Install [Grafana](https://grafana.com/grafana/download?platform=docker) for beautiful graphs.
3. By default all metrics are disabled. So specify _Kamon_ settings through _Java Properties_ and run the node
   with a desired config. For example, we ran _Graphite_ locally and it accepts _StatsD_ information on the `9999` port:

    ```bash
    SBT_OPTS="${SBT_OPTS} -Xms512M -Xmx1536M -Xss1M -XX:+CMSClassUnloadingEnabled \
    -Dkamon.modules.kamon-statsd.auto-start=yes \
    -Dkamon.modules.kamon-system-metrics.auto-start=yes \
    -Dkamon.statsd.hostname=localhost \
    -Dkamon.statsd.port=6565" sbt ytb-testnet.conf
    ```

    Here:
    * `-Dkamon.modules.kamon-statsd.auto-start=yes` enables custom metrics;
    * `-Dkamon.modules.kamon-system-metrics.auto-start=yes` enables metrics of _CPU_, _Memory_ and others;
    * See [application.conf](https://github.com/wavesplatform/Waves/blob/master/src/main/resources/application.conf)
      for more options.




# Acknowledgement

[<img src="https://www.yourkit.com/images/yklogo.png">](http://www.yourkit.com/java/profiler/index.jsp)  
We use YourKit full-featured Java Profiler to make Mir node faster. YourKit, LLC is the creator of innovative and intelligent tools for profiling Java and .NET applications.    
Take a look at YourKit's leading software products: 
<a href="http://www.yourkit.com/java/profiler/index.jsp">YourKit Java Profiler</a> and
<a href="http://www.yourkit.com/.net/profiler/index.jsp">YourKit .NET Profiler</a>.


release-notes.md
**0.6.0**

* The DEX's Order Match transaction has been changed. This is the main reason for restarting Testnet. Now, a second asset of transaction's pair is used to set an amount of the transaction.
* LPOS was implemented. New Leasing and Leasing Cancel transactions were added.
* New, HOCON based, configuration file. Old configuration file (JSON based) is supported in this release for backward compatibility. Automatic configuration file conversion added to DEB packages.

**0.3.2**

* By default walletDir and dataDir located in $HOME/ytb

**0.3.1**

* HTTP API /scorex removed. Use /node instead.

**0.2.2**

* Switch network by "testnet" in settings. Default value is true."
* /scorex/* HTTP API deprecated. Use /node/* instead.
* All logs goes to stdout and stderr. Use "loggingLevel" in config. 

**0.2.1**

* peers.dat format changed. Delete old version.
* Different HTTP status codes in replies in HTTP API were implemented
* Ytb' Scorex v1.3.2

**0.2.0**

* Peers blacklist ttl configuration via "p2p"/"blacklistResidenceTimeMilliseconds"
* Upgrade to Ytb' Scorex v1.3.1

**0.2.0-RC7**

* New API /ytb/payment returns senderPublicKey
* New API /ytb/create-signed-payment
* /ytb/external-payment deprecated. 
  Use new /ytb/broadcast-signed-payment.
* New API /ytb/payment/signature
* minimumTxFee verification for API

**0.2.0-RC5**

* /ytb/external-payment returns error for incorrect recipient

**0.2.0-RC4**

* Fixed issue with incorrect Handshake
* Balance with confirmations is the minimum balance
* /ytb/external-payment returns error if account balance invalid
* New API method /consensus/generatingbalance/{address}

**0.2.0-RC3**

* Incompatible with 0.1.3
* Upgrade to Scorex 1.2.8
* New Address format
* New hash chain for Address - Blake2b256, Keccak32
* New Testnet Genesis

**0.1.3**

* Upgrade to Scorex 1.2.6.
* New http api method /external-payment for lite client

**0.1.2**

* Upgrade to Scorex 1.2.4. Clean /scorex/ytb/data/ before run.


benchmark/src/test/scala/one/ytb/lang/v1/BigSum.scala
package one.ytb.lang.v1

import one.ytb.lang.v1.compiler.Terms.{CONST_LONG, EXPR, FUNCTION_CALL}
import one.ytb.lang.v1.evaluator.ctx.impl.PureContext
import org.openjdk.jmh.annotations.{Scope, State}

@State(Scope.Benchmark)
class BigSum {
  private val bigSum = (1 to 100).foldLeft[EXPR](CONST_LONG(0)) { (r, i) =>
    FUNCTION_CALL(
      function = PureContext.sumLong,
      args = List(r, CONST_LONG(i))
    )
  }

  val expr: EXPR = FUNCTION_CALL(
    function = PureContext.eq,
    args = List(CONST_LONG(1), bigSum)
  )
}


benchmark/src/test/scala/one/ytb/state/BaseState.scala
package one.ytb.state

import java.io.File
import java.nio.file.Files

import one.ytb.account.PrivateKeyAccount
import one.ytb.block.Block
import one.ytb.database.LevelDBWriter
import one.ytb.db.LevelDBFactory
import one.ytb.lagonaki.mocks.TestBlock
import one.ytb.mining.MiningConstraint
import one.ytb.settings.FunctionalitySettings
import one.ytb.state.diffs.BlockDiffer
import one.ytb.transaction.{GenesisTransaction, Transaction}
import org.iq80.leveldb.{DB, Options}
import org.openjdk.jmh.annotations.{Setup, TearDown}
import org.scalacheck.{Arbitrary, Gen}

trait BaseState {
  import BaseState._

  private val fsSettings: FunctionalitySettings = updateFunctionalitySettings(FunctionalitySettings.TESTNET)
  private val db: DB = {
    val dir     = Files.createTempDirectory("state-synthetic").toAbsolutePath.toString
    val options = new Options()
    options.createIfMissing(true)
    LevelDBFactory.factory.open(new File(dir), options)
  }

  val state: LevelDBWriter = new LevelDBWriter(db, fsSettings, 100000, 2000, 120 * 60 * 1000)

  private var _richAccount: PrivateKeyAccount = _
  def richAccount: PrivateKeyAccount          = _richAccount

  private var _lastBlock: Block = _
  def lastBlock: Block          = _lastBlock

  protected def mir(n: Float): Long              = (n * 100000000L).toLong
  protected val accountGen: Gen[PrivateKeyAccount] = Gen.containerOfN[Array, Byte](32, Arbitrary.arbitrary[Byte]).map(seed => PrivateKeyAccount(seed))

  protected def updateFunctionalitySettings(base: FunctionalitySettings): FunctionalitySettings = base

  protected def txGenP(sender: PrivateKeyAccount, ts: Long): Gen[Transaction]

  private def genBlock(base: Block, sender: PrivateKeyAccount): Gen[Block] =
    for {
      transferTxs <- Gen.sequence[Vector[Transaction], Transaction]((1 to TxsInBlock).map { i =>
        txGenP(sender, base.timestamp + i)
      })
    } yield
      TestBlock.create(
        time = transferTxs.last.timestamp,
        ref = base.uniqueId,
        txs = transferTxs
      )

  private val initGen: Gen[(PrivateKeyAccount, Block)] = for {
    rich <- accountGen
  } yield {
    val genesisTx = GenesisTransaction.create(rich, mir(100000000L), System.currentTimeMillis() - 10000).explicitGet()
    (rich, TestBlock.create(time = genesisTx.timestamp, Seq(genesisTx)))
  }

  protected def nextBlock(txs: Seq[Transaction]): Block = TestBlock.create(
    time = txs.last.timestamp,
    ref = lastBlock.uniqueId,
    txs = txs
  )

  private def append(prev: Option[Block], next: Block): Unit = {
    val preconditionDiff = BlockDiffer.fromBlock(fsSettings, state, prev, next, MiningConstraint.Unlimited).explicitGet()._1
    state.append(preconditionDiff, 0, next)
  }

  def applyBlock(b: Block): Unit = {
    append(Some(lastBlock), b)
    _lastBlock = b
  }

  def genAndApplyNextBlock(): Unit = {
    val b = genBlock(lastBlock, richAccount).sample.get
    applyBlock(b)
  }

  @Setup
  def init(): Unit = {
    val (richAccount, genesisBlock) = initGen.sample.get
    _richAccount = richAccount

    append(None, genesisBlock)
    _lastBlock = genesisBlock
  }

  @TearDown
  def close(): Unit = {
    db.close()
  }
}

object BaseState {
  private val TxsInBlock = 5000
}


benchmark/src/test/scala/one/ytb/lang/v1/CompilerBenchmark.scala
package one.ytb.lang.v1

import java.util.concurrent.TimeUnit

import one.ytb.lang.v1.CompilerBenchmark.St
import org.openjdk.jmh.annotations._
import org.openjdk.jmh.infra.Blackhole
import one.ytb.transaction.smart.script.ScriptCompiler

@OutputTimeUnit(TimeUnit.NANOSECONDS)
@BenchmarkMode(Array(Mode.AverageTime))
@Threads(1)
@Fork(1)
@Warmup(iterations = 10)
@Measurement(iterations = 10)
class CompilerBenchmark {

  @Benchmark
  def serialize_test(st: St, bh: Blackhole): Unit = bh.consume(ScriptCompiler(st.scriptString, isAssetScript = false).explicitGet()._1)
}

object CompilerBenchmark {

  @State(Scope.Benchmark)
  class St {
    val scriptString: String =
      """
        |let alicePubKey  = base58'B1Yz7fH1bJ2gVDjyJnuyKNTdMFARkKEpV'
        |let bobPubKey    = base58'7hghYeWtiekfebgAcuCg9ai2NXbRreNzc'
        |let cooperPubKey = base58'BVqYXrapgJP9atQccdBPAgJPwHDKkh6A8'
        |
        |let aliceSigned  = if(sigVerify(tx.bodyBytes, tx.proofs[0], alicePubKey  )) then 1 else 0
        |let bobSigned    = if(sigVerify(tx.bodyBytes, tx.proofs[1], bobPubKey    )) then 1 else 0
        |let cooperSigned = if(sigVerify(tx.bodyBytes, tx.proofs[2], cooperPubKey )) then 1 else 0
        |
        |aliceSigned + bobSigned + cooperSigned >= 2
      """.stripMargin
  }
}


benchmark/src/main/scala/one/ytb/state/Settings.scala
package one.ytb.state

import com.typesafe.config.Config
import net.ceedubs.ficus.Ficus._
import net.ceedubs.ficus.readers.ArbitraryTypeReader._

case class Settings(networkConfigFile: String,
                    aliasesFile: String,
                    aliasesFromHeight: Int,
                    restTxsFile: String,
                    restTxsFromHeight: Int,
                    accountsFile: String,
                    accountsFromHeight: Int,
                    assetsFile: String,
                    assetsFromHeight: Int,
                    dataFile: String,
                    dataFromHeight: Int)

object Settings {
  def fromConfig(config: Config): Settings = {
    implicit val _ = net.ceedubs.ficus.readers.namemappers.HyphenNameMapper
    config.as[Settings]("ytb.benchmark.state")
  }
}


benchmark/src/main/scala/one/ytb/state/ExtractInfo.scala
package one.ytb.state

import java.io.{File, PrintWriter}
import java.util.concurrent.ThreadLocalRandom

import com.typesafe.config.ConfigFactory
import one.ytb.database.LevelDBWriter
import one.ytb.db.LevelDBFactory
import one.ytb.lang.v1.traits.DataType
import one.ytb.settings.{YtbSettings, loadConfig}
import one.ytb.state.bench.DataTestData
import org.iq80.leveldb.{DB, Options}
import scodec.bits.{BitVector, ByteVector}
import one.ytb.account.AddressScheme
import one.ytb.utils.ScorexLogging
import one.ytb.block.Block
import one.ytb.transaction.assets.IssueTransaction
import one.ytb.transaction.{Authorized, CreateAliasTransactionV1, DataTransaction, Transaction}

import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.util.control.NonFatal

/**
  * Extracts data from the database to use it in RealDbBenchmark.
  * Requires a separate main file because takes too long time to run.
  */
object ExtractInfo extends App with ScorexLogging {

  if (args.length < 1) {
    log.error("Specify a path to the node config. Usage: benchmark/run /full/path/to/the/config.conf")
    System.exit(1)
  }

  val benchSettings = Settings.fromConfig(ConfigFactory.load())
  val ytbSettings = {
    val config = loadConfig(ConfigFactory.parseFile(new File(args.head)))
    YtbSettings.fromConfig(config)
  }

  AddressScheme.current = new AddressScheme {
    override val chainId: Byte = ytbSettings.blockchainSettings.addressSchemeCharacter.toByte
  }

  val db: DB = {
    val dir = new File(ytbSettings.dataDirectory)
    if (!dir.isDirectory) throw new IllegalArgumentException(s"Can't find directory at '${ytbSettings.dataDirectory}'")
    LevelDBFactory.factory.open(dir, new Options)
  }

  try {
    val state = new LevelDBWriter(db, ytbSettings.blockchainSettings.functionalitySettings, 100000, 2000, 120 * 60 * 1000)

    def nonEmptyBlockHeights(from: Int): Iterator[Integer] =
      for {
        height     <- randomInts(from, state.height)
        (block, _) <- state.blockHeaderAndSize(height)
        if block.transactionCount > 0
      } yield height

    def nonEmptyBlocks(from: Int): Iterator[Block] =
      nonEmptyBlockHeights(from)
        .flatMap(state.blockAt(_))

    val aliasTxs = nonEmptyBlocks(benchSettings.aliasesFromHeight)
      .flatMap(_.transactionData)
      .collect {
        case _: CreateAliasTransactionV1 => true
      }

    val restTxs = nonEmptyBlocks(benchSettings.restTxsFromHeight)
      .flatMap(_.transactionData)

    val accounts = for {
      b <- nonEmptyBlocks(benchSettings.accountsFromHeight)
      sender <- b.transactionData
        .collect {
          case tx: Transaction with Authorized => tx.sender
        }
        .take(100)
    } yield sender.toAddress.stringRepr
    write("accounts", benchSettings.accountsFile, takeUniq(1000, accounts))

    val aliasTxIds = aliasTxs.map(_.asInstanceOf[CreateAliasTransactionV1].alias.stringRepr)
    write("aliases", benchSettings.aliasesFile, aliasTxIds.take(1000))

    val restTxIds = restTxs.map(_.id().base58)
    write("rest transactions", benchSettings.restTxsFile, restTxIds.take(10000))

    val assets = nonEmptyBlocks(benchSettings.assetsFromHeight)
      .flatMap { b =>
        b.transactionData.collect {
          case tx: IssueTransaction => tx.assetId()
        }
      }
      .map(_.base58)

    write("assets", benchSettings.assetsFile, takeUniq(300, assets))

    val data = for {
      b <- nonEmptyBlocks(benchSettings.dataFromHeight)
      test <- b.transactionData
        .collect {
          case tx: DataTransaction =>
            val addr = ByteVector(tx.sender.toAddress.bytes.arr)
            tx.data.collectFirst {
              case x: IntegerDataEntry => DataTestData(addr, x.key, DataType.Long)
              case x: BooleanDataEntry => DataTestData(addr, x.key, DataType.Boolean)
              case x: BinaryDataEntry  => DataTestData(addr, x.key, DataType.ByteArray)
              case x: StringDataEntry  => DataTestData(addr, x.key, DataType.String)
            }
        }
        .take(50)
      r <- test
    } yield {
      val x: BitVector = DataTestData.codec.encode(r).require
      x.toBase64
    }
    write("data", benchSettings.dataFile, data.take(400))
  } catch {
    case NonFatal(e) => log.error(e.getMessage, e)
  } finally {
    db.close()
    log.info("Done")
  }

  def takeUniq[T](size: Int, xs: Iterator[T]): mutable.Set[T] = {
    val r = mutable.Set.empty[T]
    xs.find { x =>
      r.add(x)
      r.size == size
    }
    r
  }

  def write(label: String, absolutePath: String, data: TraversableOnce[String]): Unit = {
    log.info(s"Writing $label to '$absolutePath'")
    val printWriter = new PrintWriter(absolutePath)
    data.foreach(printWriter.println)
    printWriter.close()
  }

  def randomInts(from: Int, to: Int): Iterator[Integer] =
    ThreadLocalRandom
      .current()
      .ints(from, to)
      .iterator()
      .asScala

}
